hand-eye-calibration/full_hand_eye_sample.py

```diff
@@ -0,0 +1,486 @@
+    """
+
+    if np.isnan(angle):
+        angle = np.linalg.norm(rotvec)
+
+    quat = np.zeros((4))
+    quat[0] = np.cos(0.5 * angle)
+    q_r = np.sqrt((1 - np.power(quat[0], 2)) / np.power(np.linalg.norm(rotvec), 2))
+    quat[1] = q_r * rotvec[0]
+    quat[2] = q_r * rotvec[1]
+    quat[3] = q_r * rotvec[2]
+    quat = quat / np.sign(quat[0])
+
+    return quat
+
+
+def _calculate_transform_matrix(
+    robot_pose_before: np.array, robot_pose_after: np.array
+):
+    """Compute transform matrix based on poses
+
+    Args:
+        robot_pose_before: Pose of robot taken right before image was captured
+        robot_pose_after: Pose of robot right after image was captured
+
+    Returns:
+        transform: 4x4 Transform matrix
+    """
+
+    pose = (robot_pose_before + robot_pose_after) / 2
+    translation = pose[:3] * 1000
+    rotvec = pose[3:]
+    tansform = np.eye(4)
+    tansform[:3, :3] = quat_to_rotm(rotvec_to_quat(rotvec))
+    tansform[:3, 3] = translation.T
+    return tansform
+
+
+def _write_robot_state(
+    con: rtde, input_data, finish_capture: bool = False, camera_ready: bool = False
+):
+    """Write to robot I/O registrer
+
+    Args:
+        con: Connection between computer and robot
+        input_data: Input package containing the specific input data registers
+        finish_capture: Boolean value to robot_state that q_r scene capture is finished
+        camera_ready: Boolean value to robot_state that camera is ready to capture images
+    """
+
+    if finish_capture:
+        input_data.input_bit_register_64 = 1
+    else:
+        input_data.input_bit_register_64 = 0
+
+    if camera_ready:
+        input_data.input_bit_register_65 = 1
+    else:
+        input_data.input_bit_register_65 = 0
+
+    con.send(input_data)
+
+
+def _initialize_robot_sync(host: str, port: int):
+    """Set up communication with UR robot
+
+    Args:
+        host: IP address to host
+        port: Port number
+
+    Returns:
+        con: Connection to robot
+        input_data: Input package containing the specific input data registers
+
+    Raises:
+        Exception: If computer is not able to establish comminucation with robot
+    """
+
+    conf = rtde_config.ConfigFile(Path(Path.cwd() / "robot_communication_file.xml"))
+    output_names, output_types = conf.get_recipe("out")
+    input_names, input_types = conf.get_recipe("in")
+
+    # Rtde communication is by default port 30004.
+    con = rtde.RTDE(host, port)
+    con.connect()
+
+    # To ensure that the application and further versions of UR controller is compatible
+    if not con.negotiate_protocol_version():
+        raise Exception(f"Protocol do not match")
+
+    if not con.send_output_setup(output_names, output_types, frequency=200):
+        raise Exception(f"Unable to configure output")
+
+    input_data = con.send_input_setup(input_names, input_types)
+
+    if not con.send_start():
+        raise Exception(f"Unable to start synchronization")
```

Comment:
We usually don't use the Exception base-class. Use something more specific, like `RuntimeError()`

Reply:
Done